Design/histogram_type.md

```diff
@@ -0,0 +1,251 @@
+- Enable unit testing of key concepts in Mantid.
+  - Arithmetic operations with histograms.
+  - Propagation of errors.
+- Enforce type safety.
+- Do not compromise on speed.
+
+
+
+## Implementation
+
+#### Basics
+
+The basic concept seems simple enough:
+
+- Add one class for a histogram with an interface that provides standard operations for histogram data and point data.
+  - Note that we do not introduce a separate type for histogram and distribution data. This distinction would be only internal, and access to either histogram or distribution data is provided via public interface methods. The goal is to eliminate this distinction from as much of the code as possible.
+  - For histogram and point data the argument is not so clear. It is not possible to do a unique conversion from point data to histogram data, so the internal storage type is not arbitrary here, and conversion are lossy. Nevertheless, considering how many of our current algorithms deal with both types of data in the same way, it is probably a good idea to also keep both histogram and point data in the same type. The interface will have to reflect the internal differences in this case and cannot be symmetric as in the distribution-data case.
+- Add a few classes for `X`, `Y`, and `E` (etc.) that would mainly be used internally in the histogram class and for direct access to the underlying data.
+
+#### Examples
+
+Before going into further details it is best to have a look at the envisioned end result.
+
+We introduce types for all the different interpretations of a `std::vector<double>` in this context.
+The new types would be simple wrappers around `std::vector<double>` (or rather `Kernel::cow_ptr<std::vector<double>>`), and add a suitable interface to it.
+In particular:
+
+1. X-data: We need to cover two cases and introduce two types: `BinEdges` and `Points`.
+2. Y-data: We need to cover two cases and introduce two types: `Counts` and `Frequencies`.
+3. Errors: All four types introduced by 1. and 2. can come with uncertainties.
+   They can be represented as variances or standard deviations.
+   We thus introduce 8 types: `BinEdgeVariances`, `BinEdgeStandardDeviations`, `PointVariances`, `PointStandardDeviations`, `CountVariances`, `CountStandardDeviation`, `FrequencyVariances`, `FrequencyStandardDeviations`.
+
+The amount of new types introduced may seem appalling at first, but we usually do not have to deal with many of them:
+
+1. For 'standard' operations you do not need to work with the low-level types, but can use the interface of `Histogram`, e.g.,
+
+   ```cpp
+   Histogram hist1;
+   Histogram hist2;
+   // Initialize histograms
+   // ...
+   // Adding histograms can automatically add `Y`, and propagates the errors properly
+   auto result = hist1 + hist2;
+   ```
+
+   Details of the `Histogram` interface are still open, but it should be simple to extend it once the roll-out is done.
```

Comment:
I also think that automatic rebinning should not be performed. I'd be cautious about doing anything automatic or unexpected via the binary operators. Explicit convenience methods to rebin and add would be fine in addition to `operator+`.